pkg/backends/calico/routes_test.go

```diff
@@ -110,88 +132,120 @@ var _ = Describe("RouteGenerator", func() {
+				Expect(rg.client.cacheRevision).To(Equal(initRevision + 2))
+				Expect(rg.svcRouteMap["foo/bar"]).To(Equal(expectedSvcRouteMap))
+				Expect(rg.routeAdvertisementCount["127.0.0.1/32"]).To(Equal(1))
+				Expect(rg.routeAdvertisementCount["172.217.3.5/32"]).To(Equal(1))
 				Expect(rg.client.cache["/calico/staticroutes/127.0.0.1-32"]).To(Equal("127.0.0.1/32"))
+				Expect(rg.client.cache["/calico/staticroutes/172.217.3.5-32"]).To(Equal("172.217.3.5/32"))
 
 				// set to unsupport service type
 				svc.Spec.Type = v1.ServiceTypeExternalName
 				rg.onSvcUpdate(nil, svc)
-				Expect(rg.client.cacheRevision).To(Equal(initRevision + 2))
-				Expect(rg.svcRouteMap).ToNot(HaveKey("foo/bar"))
+				Expect(rg.client.cacheRevision).To(Equal(initRevision + 4))
+				Expect(rg.svcRouteMap["foo/bar"]).ToNot(HaveKey("172.217.3.5/32"))
+				Expect(rg.svcRouteMap["foo/bar"]).ToNot(HaveKey("127.0.0.1-32"))
+				Expect(rg.routeAdvertisementCount["127.0.0.1/32"]).To(Equal(0))
+				Expect(rg.routeAdvertisementCount["172.217.3.5/32"]).To(Equal(0))
+				Expect(rg.client.cache).ToNot(HaveKey("/calico/staticroutes/172.217.3.5-32"))
 				Expect(rg.client.cache).ToNot(HaveKey("/calico/staticroutes/127.0.0.1-32"))
 			})
 		})
 
 		Context("onEp[Add|Delete]", func() {
-			It("should add the service's clusterIP into the svcRouteMap", func() {
+			It("should add the service's cluster IP and whitelisted external IPs into the svcRouteMap", func() {
 				// add
 				initRevision := rg.client.cacheRevision
 				rg.onEPAdd(ep)
-				Expect(rg.client.cacheRevision).To(Equal(initRevision + 1))
-				Expect(rg.svcRouteMap["foo/bar"]).To(Equal("127.0.0.1/32"))
+				Expect(rg.client.cacheRevision).To(Equal(initRevision + 2))
+				Expect(rg.svcRouteMap["foo/bar"]).To(Equal(expectedSvcRouteMap))
+				Expect(rg.routeAdvertisementCount["127.0.0.1/32"]).To(Equal(1))
+				Expect(rg.routeAdvertisementCount["172.217.3.5/32"]).To(Equal(1))
 				Expect(rg.client.cache["/calico/staticroutes/127.0.0.1-32"]).To(Equal("127.0.0.1/32"))
+				Expect(rg.client.cache["/calico/staticroutes/172.217.3.5-32"]).To(Equal("172.217.3.5/32"))
 
 				// delete
 				rg.onEPDelete(ep)
-				Expect(rg.client.cacheRevision).To(Equal(initRevision + 2))
+				Expect(rg.client.cacheRevision).To(Equal(initRevision + 4))
 				Expect(rg.svcRouteMap).ToNot(HaveKey("foo/bar"))
+				Expect(rg.routeAdvertisementCount["127.0.0.1/32"]).To(Equal(0))
+				Expect(rg.routeAdvertisementCount["172.217.3.5/32"]).To(Equal(0))
 				Expect(rg.client.cache).ToNot(HaveKey("/calico/staticroutes/127.0.0.1-32"))
+				Expect(rg.client.cache).ToNot(HaveKey("/calico/staticroutes/172.217.3.5-32"))
 			})
 		})
 
 		Context("onEpDelete", func() {
-			It("should add the service's clusterIP into the svcRouteMap and then remove it for unsupported service type", func() {
+			It("should add the service's cluster IP and whitelisted external IPs into the svcRouteMap and then remove it for unsupported service type", func() {
```

Comment:
I think it's worth adding a new test case specifically for the scenario we were talking about above with two services triggering advertisement of the same route, then withdrawing one and making sure we don't with draw the route, then withdrawing the other and making sure we do withdraw the route.